### PR TITLE
Mamba and S6

### PR DESCRIPTION
**Pull request by Benjamin-Walker**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,6 +25,56 @@ want to predict.
...
+
+This repository is implemented in python 3.10 and most of the experiments use Jax as their machine learning framework. 
+However, in order to use the efficient Pytorch implementation of the [Mamba recurrence](https://github.com/state-spaces/mamba/tree/main), 
+the code for S6 and Mamba is implemented in Pytorch. Although it is possible to install the cuda versions of Jax and 
+Pytorch in the same environment, we recommend using two separate environments. The repository is designed such that the
+Jax environment is the main environment and the Pytorch environment is only used for the S6 and Mamba experiments.
+
+### Jax Environment
+
+The code for S5, LRU, NCDE, NRDE, and Log-NCDE uses the following packages:
```
Comment (Benjamin-Walker) [inline README.md:38]: `The code for preprocessing the datasets, training S5, LRU, NCDE, and Log-NCDE, and plotting the results uses the following packages:'

File: torch_experiments/train.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            shutil.rmtree(output_dir)
+            os.makedirs(output_dir)
+            print(f"Directory {output_dir} has been deleted and recreated.")
+        else:
+            raise ValueError(f"Directory {output_dir} already exists. Exiting.")
+    else:
+        os.makedirs(output_dir)
+        print(f"Directory {output_dir} has been created.")
+
+    n_samples = {
```
Comment (Benjamin-Walker) [inline torch_experiments/train.py:138]: Change this to be an argument of create_dataset_model_and_train

File: torch_experiments/train.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        "SelfRegulationSCP2": 380,
+        "ppg": 1232,
+        "signature1": 100000,
+        "signature2": 100000,
+        "signature3": 100000,
+        "signature4": 100000,
+    }
+    indexes = torch.randperm(n_samples[dataset_name])
+
+    device = "cuda"
```
Comment (Benjamin-Walker) [inline torch_experiments/train.py:153]: Shouldn't be hardcoded

File: torch_experiments/train.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    ).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
+    running_loss = 0.0
+    all_train_accuracies = []
+    all_val_accuracies = []
+    val_acc_for_best_model = []
+    no_val_improvement = 0.0
+    steps = []
+    step = 0
+    num_epochs = 10 * int(num_steps / len(train_dataloader))
```
Comment (Benjamin-Walker) [inline torch_experiments/train.py:219]: Not the prettiest approach

File: torch_experiments/train.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        input_dim=input_dim,
+        output_dim=output_dim,
+        classification=classification,
+        S6=S6,
+        output_step=output_step,
+        **model_args,
+    ).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
+    running_loss = 0.0
+    all_train_accuracies = []
```
Comment (Benjamin-Walker) [inline torch_experiments/train.py:213]: Change accuracy to metric throughout

File: torch_experiments/train.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+                ):
+                    no_val_improvement += 1
+                    if no_val_improvement > 10:
+                        steps_save = np.array(steps)
+                        all_train_accuracies_save = np.array(all_train_accuracies)
+                        all_val_accuracies_save = np.array(all_val_accuracies)
+                        test_accuracy = np.array(test_accuracy)
+                        np.save(output_dir + "/steps.npy", steps_save)
+                        np.save(
+                            output_dir + "/train_acc.npy", all_train_accuracies_save
```
Comment (Benjamin-Walker) [inline torch_experiments/train.py:315]: `all_train_acc.npy`

**Review by Benjamin-Walker — COMMENTED** [review]
Few small things to fix before merging
